pkg/resources/catalog.cattle.io/v1/clusterrepo/validator_test.go

```diff
@@ -0,0 +1,119 @@
+package clusterrepo
+
+import (
+	"context"
+	"encoding/json"
+	"testing"
+
+	catalogv1 "github.com/rancher/rancher/pkg/apis/catalog.cattle.io/v1"
+	"github.com/rancher/webhook/pkg/admission"
+	"github.com/stretchr/testify/assert"
+	admissionv1 "k8s.io/api/admission/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func TestClusterRepoValidation(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name        string
+		clusterRepo *catalogv1.ClusterRepo
+		operation   admissionv1.Operation
+		wantAllowed bool
+		wantErr     bool
+	}{
+		{
+			name: "Both GitRepo and URL are set when creating",
+			clusterRepo: &catalogv1.ClusterRepo{
+				Spec: catalogv1.RepoSpec{
+					URL:     "https://url.com",
+					GitRepo: "https://github.com",
+				},
+			},
+			operation:   admissionv1.Create,
+			wantAllowed: false,
+			wantErr:     true,
+		},
+		{
+			name: "Both GitRepo and URL are set when updating",
+			clusterRepo: &catalogv1.ClusterRepo{
+				Spec: catalogv1.RepoSpec{
+					URL:     "https://url.com",
+					GitRepo: "https://github.com",
+				},
+			},
+			operation:   admissionv1.Create,
+			wantAllowed: false,
+			wantErr:     true,
+		},
+		{
+			name: "Only GitRepo is set",
+			clusterRepo: &catalogv1.ClusterRepo{
+				Spec: catalogv1.RepoSpec{
+					GitRepo: "https://github.com",
+				},
+			},
+			operation:   admissionv1.Update,
+			wantAllowed: true,
+			wantErr:     false,
+		},
+		{
+			name: "Only URL is set",
+			clusterRepo: &catalogv1.ClusterRepo{
+				Spec: catalogv1.RepoSpec{
+					URL: "https://url.com",
+				},
+			},
+			operation:   admissionv1.Update,
+			wantAllowed: true,
+			wantErr:     false,
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
```

Comment:
You can run subtests in parallel, too.
```suggestion
		t.Run(test.name, func(t *testing.T) {
		      t.Parallel()
```

Reply:
Done.